common/src/stack/command/stack/commands/set/host/power/__init__.py

```diff
@@ -8,96 +8,139 @@
-			stderr = subprocess.STDOUT)
-		out, err = p.communicate()
-
-		if self.debug:
-			self.beginOutput()
-			self.addOutput(host, out.decode())
-			self.endOutput(padChar='', trimOwner=True)
+	<example cmd='set host power stacki-1-1 command=off use-method=ssh'>
+	Turns off host stacki-1-1 using ssh.
+	</example>
+	"""
+	def mq_publish(self, host, cmd):
+		"""
+		Publish the power status to the
+		message queue
+		"""
 
 		ttl = 60*10
 		if cmd == 'off':
 			ttl = -1
 
-		msg = { 'source' : host, 
-			'channel': 'health', 
+		msg = { 'source' : host,
+			'channel': 'health',
 			'ttl'    : ttl,
 			'payload': '{"state": "power %s"}' % cmd }
 
 		tx = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
-		tx.sendto(json.dumps(msg).encode(), 
+		tx.sendto(json.dumps(msg).encode(),
 			  ('127.0.0.1', stack.mq.ports.publish))
 		tx.close()
 
-
-
 	def run(self, params, args):
 		if not len(args):
 			raise ArgRequired(self, 'host')
 
-		cmd, debug = self.fillParams([ ('command', None, True), ('debug', False) ])
-
-		if cmd == 'status':
-			#
-			# used by "stack list host power" -- this is an easy way in which to
-			# share code between the two commands
-			#
-			# set 'debug' to True in order to get output from the status command
-			#
-			debug = True
-		elif cmd not in [ 'on', 'off', 'reset' ]:
-			raise ParamError(self, 'command', 'must be "on", "off" or "reset"')
-		
+		cmd, debug, force_imp = self.fillParams([
+			('command', None, True),
+			('debug', False),
+			('use-method', None)
+		])
+		imp_names = []
 		self.debug = self.str2bool(debug)
-		
+
+		# Flag for if an implementation has
+		# successfully run
+		imp_success = False
+
+		self.loadImplementation()
+
+		# The first implementation is ipmi by default
+		# but can be forced by the use-method parameter
+		# If this is set, only that implementation will be run
+		if force_imp and force_imp in self.impl_list:
+			imp_names.append(force_imp)
+
+		# Otherwise use all the implementations
+		else:
+			# Gather all set power implmentations besides ipmi and ssh
+			imp_names  = [imp for imp in self.impl_list if 'ssh' not in imp and 'ipmi' not in imp]
+
+			# Add ipmi to be the first implementation used
+			if 'ipmi' in self.impl_list:
+				imp_names[:0] = ['ipmi']
+
+			# Add ssh to be the last implementation used
+			if 'ssh' in self.impl_list:
+				imp_names.append('ssh')
+
+		if cmd not in [ 'on', 'off', 'reset', 'status' ]:
+			raise ParamError(self, 'command', 'must be "on", "off", "reset", "status"')
+
 		for host in self.getHostnames(args):
-			for o in self.call('list.host.interface', [ host ]):
-				if o['interface'] == 'ipmi':
-					self.doPower(host, o['ip'], cmd)
+			debug_msgs = []
+
+			# msgs gathers the normal output of the command
+			msgs = []
+			self.beginOutput()
+
+			for imp in imp_names:
+				debug_msgs.append(f'Attempting to set power via {imp}')
+				try:
+					imp_msg = self.runImplementation(imp, [host, cmd])
+					if imp_msg:
+						msgs.append(str(imp_msg))
```

Comment:
What's this `str()` for?  Aren't the imps always returning a string?  See below as well.

Reply:
This accomplishes the same as `print(Exception)` in that it gets the exception text as a string instead of an Exception object.